buildpack/telemetry/dynatrace.py

```diff
@@ -5,23 +5,165 @@
+    except Exception as e:
+        logging.warning(
+            "Failed to parse Dynatrace manifest file", exc_info=True
+        )
+        return
+
+    # dynamic default
+    default_env.update({"DT_TENANTTOKEN": manifest.get("tenantToken")})
+
+    for key, dv in default_env.items():
+        value = os.environ.get(key, dv)
+        if value:
+            util.upsert_custom_environment_variable(m2ee, key, value)
+    util.upsert_custom_environment_variable(
+        m2ee, "DT_CONNECTION_POINT", get_connection_endpoint()
+    )
+
+    agent_path = os.path.join(".local", get_agent_path())
+    if not os.path.exists(agent_path):
+        raise Exception(
+            "Dynatrace Agent not found: {agent_path}".format(agent_path)
+        )
+
+    util.upsert_javaopts(
+        m2ee,
+        [
+            "-agentpath:{path}".format(path=os.path.abspath(agent_path)),
+            "-Xshare:off",
+        ],
+    )
+
+
+def get_manifest():
+    with open(".local/manifest.json", "r") as f:
+        return json.load(f)
+
+
+def get_connection_endpoint():
+    manifest = get_manifest()
+    endpoints = manifest.get("communicationEndpoints", [])
+    # prepend the DT_SAAS_URL because the communication endpoints might not be correct
+    endpoints.insert(
+        0, "{url}/communication".format(url=os.environ.get("DT_SAAS_URL"))
+    )
+    return ";".join(endpoints)
+
+
+def get_agent_path():
+    manifest = get_manifest()
+    technologies = manifest.get("technologies")
+    java_binaries = technologies.get("java").get("linux-x86-64")
+    for f in java_binaries:
+        binary_type = f.get("binarytype")
+        if binary_type == "loader":
+            return f.get("path")
+
+
+def is_telegraf_enabled():
     return (
         "DT_PAAS_TOKEN" in os.environ.keys()
         and "DT_SAAS_URL" in os.environ.keys()
     )
 
 
+def is_agent_enabled():
+    return is_telegraf_enabled() and ("DT_TENANT" in os.environ.keys())
+
+
 def get_ingestion_info():
-    if not is_enabled():
+    if not is_telegraf_enabled():
         return None, None
 
-    logging.info("Metrics ingestion to Dynatrace is configured")
+    logging.info("Metrics ingestion to Dynatrace via telegraf is configured")
     token = os.getenv("DT_PAAS_TOKEN")
-    ingest_url = urljoin(os.getenv("DT_SAAS_URL"), INGEST_ENDPOINT)
+    ingest_url = _get_ingestion_url(os.getenv("DT_SAAS_URL"), INGEST_ENDPOINT)
     return token, ingest_url
+
+
+def _get_ingestion_url(saas_url, endpoint):
+    """
+    Basic url join but purposefully isolated to add some unittests easily.
+    When merging an url and an additional endpoint, python's urljoin method
+    has so many little details. See:
+    https://stackoverflow.com/questions/10893374/python-confusions-with-urljoin
+
+    So, basically we need to make sure that the url ends with '/' and
+    the endpoint does not start with '/'
+    """
+
+    saas_url = f"{saas_url}/"
+    if endpoint.startswith('/'):
+        endpoint = endpoint[1:]
```

Comment:
This endpoint is a constant and hardcoded, so isn't this check a bit redundant? 
For future scenarios, the unit tests should flag any issues.

And why not refer to the constant directly 🤔 It also simplifies the unit test.

Reply:
I wanted to isolate the behavior regardless of the value of the constant. Basically, if I use the constant here, I won't be able to test its different values.

Reply:
True, injecting vars makes the tests easier. So maybe have the param but default to the top-level constant and do not have the check as the unittests should flag any issues right? Would that work?

Reply:
Changed to lstrip, still behaviour is the same. Just wanna be sure that this method is future proof. If constant changes to have a "/" in the beginning don't affect anything when it's like this